Generate changelog for 2.11.6 (#11188)
tarantool/tarantool: changelogs/2.11.6.md

# 2.11.6

Date: 2025-02-24
Tag: 2.11.6

## Announcement: `double` field type problems

Users who ever used the `double` field type in vinyl indexes might have their
indexes broken. Regardless of this release. Preferably prior to the upgrade the
users must get rid of the `double` indexed field type in vinyl. It must be
drop-in replaceable by the `number` field type.

The reason is that the sorting of `double` fields would always compare the field
values as C-language doubles, even when actual integers are stored. This works
fine for small numbers, but starting from 2^53 the double-style comparison

> looses precision. For example, 18446744073709551615 and 18446744073709551614

Reviewer: ```suggestion
loses precision. For example, 18446744073709551615 and 18446744073709551614
```

Author: Fixed.

Author: Sent for `3.4.0`, `3.3.2`, `3.2.2` in #11196.

would be considered the same values.

Besides, point-lookups in such index are going to be broken by design too -
looking up by 18446744073709551615 can return 18446744073709551614 and many
other integers around this point.

Unfortunately, even if the user doesn't have the `double` type in any vinyl
indexes, but **ever had it before**, it is still unsafe. A `double` index
altered to `number` or `scalar` would remain broken. It is **very important that

> the users rebuild all the vinyl indexes that have or ever had the `double` field

Reviewer: ```suggestion
users rebuild all the vinyl indexes that have or ever had the `double` field
```

Author: Fixed.

Author: Sent for `3.4.0`, `3.3.2`, `3.2.2` in #11196.

type in them**. What is worse, just an alter `double` -> `number` **won't
work**. The old index must be dropped and a new one must be created. In any
order.

See more details in gh-9965 and [gh-11017][gh-11017].

[gh-11017]: https://github.com/orgs/tarantool/discussions/11017

## Overview

2.11.6 is the next stable release in the [long-term support (LTS)

> version][release_policy] 2.11.x release series.

Reviewer: ```suggestion
version][release_policy] of the 2.11.x release series.
```

Author: Fixed.

Author: At the end, I rewrote the release notes template almost entirely, see #11210.

  allowed to insert the same key multiple times into a unique index, and
  sometimes wouldn't allow to find an existing key in an index. Could happen
  when numbers were encoded in MessagePack suboptimally and when the `double`
  field type was used (gh-9965).
* Fixed the issue with hanging write operations forever triggered by heavy
  write load and WAL writing failures on cascade rollback (gh-11081).

### Memtx

* Fixed a crash when memtx MVCC tried to abort an already committed
  non-memtx transaction if it used a system space view or tried to
  perform a DDL operation (gh-10715).
* Fixed a bug when Tarantool with memtx MVCC enabled was aborted on
  workload with many `index:get()` operations reading nothing (gh-11022).

### Vinyl

* Fixed a bug when a deleted tuple wasn't purged from a secondary index data
  stored on disk in case it was updated more than once in the same transaction.
  The bug couldn't result in inconsistent query results, but it could lead to
  performance degradation and increased disk usage (gh-10820, gh-10822).
* Fixed a bug when joining a new replica to a master instance that experiences

>   a heavy write load would severely degrade the master instance performance.

Reviewer: ```suggestion
  a heavy write load severely degrades the master instance performance.
```

Author: Fixed.

Author: Sent for `3.2.2` in #11197. Can't send it for 3.3.2 and 3.4.0, because the change is already released in 3.3.0.

  The fix should also speed up long-running scan requests (gh-10846).
* Fixed a bug when a tuple could disappear from a multikey index in case it
  replaced a tuple with duplicate multikey array entries created in the same
  transaction. With the enabled `defer_deletes` space option, the bug could
  also trigger a crash (gh-10869, gh-10870).
* Fixed a bug when the tuple cache was not properly invalidated in case
  a WAL write error occurred while committing a `space.delete()` operation.
  The bug could lead to a crash or an invalid read query result (gh-10879).
* Fixed a bug when a deleted secondary index key wasn't purged on compaction
  of a space with the `defer_deletes` option enabled (gh-10895).
* Fixed a bug in the tuple cache when a transaction operating in a read view
  could skip a tuple deleted after the read view creation (gh-11079).
* Fixed a bug in the tuple cache when a tuple could become inaccessible via
  a secondary index after a transaction rollback caused by a WAL write error
  (gh-11140).

### Replication

* RAFT requests are now sent with 0 (not local) GROUP_ID (gh-10727).

> * Fixed a bug when the Raft state wasn't sent during the `META_JOIN` stage,

Reviewer: ```suggestion
* Fixed a bug when the RAFT state wasn't sent during the `META_JOIN` stage,
```

Author: Fixed.

Author: Sent for `3.2.2` in #11197. Can't send it for 3.3.2 and 3.4.0, because the change is already released in 3.3.0.
